src/core/service.c

```diff
@@ -2144,6 +2145,71 @@ _pure_ static bool service_can_reload(Unit *u) {
         return !!s->exec_command[SERVICE_EXEC_RELOAD];
 }
 
+static int service_exec_command_index(Unit *u, ServiceExecCommand id, bool control) {
+        Service *s = SERVICE(u);
+        unsigned index = 0;
+        ExecCommand *current, *first, *c;
+
+        assert(s);
+
+        first = s->exec_command[id];
+        current = control ? s->control_command : s->main_command;
+
+        /* Figure out where we are in the list by walking back to the beginning */
+        for (c = current; c != first; c = c->command_prev)
+                index++;
+
+        return index;
+}
+
+static int service_serialize_exec_command(Unit *u, FILE *f, bool control) {
+        Service *s = SERVICE(u);
+        ServiceExecCommand id;
+        ExecCommand *c;
+        int index, r;
+        const char *type;
+        char **arg;
+        _cleanup_strv_free_ char **escaped_args = NULL;
+        _cleanup_free_ char *args = NULL;
+
+        assert(s);
+        assert(f);
+
+        if (control) {
+                type = "control";
+                c = s->control_command;
+                id = s->control_command_id;
+        } else {
+                type = "main";
+                c = s->main_command;
+                id = SERVICE_EXEC_START;
+        }
+
+        assert(c);
+
+        index = service_exec_command_index(u, id, control);
+
+        STRV_FOREACH(arg, c->argv) {
+                char *e = NULL;
+
+                e = cescape(*arg);
+                if (!e)
+                        return -ENOMEM;
+
+                r = strv_push(&escaped_args, e);
+                if (r < 0)
+                        return -ENOMEM;
+        }
+
+        args = strv_join(escaped_args, " ");
+        if (!args)
+                return -ENOMEM;
```

Comment:
hmm, I think instead of preparing a list and then calling strv_join() and throwing the list away again, it would be better to just prepare the string directly. something like this:

```c
…
size_t length = 0, allocated = 0;
char *args = NULL;
…
STRV_FOREACH(arg, c->argv) {
        size_t n;
        char *e;

        e = cescape(arg);
        if (!e)
                return -ENOMEM;

        n = strlen(e);
        if (!GREEDY_REALLOC(args, allocated, length + 1 + n + 1))
                return -ENOMEM;

        if (length > 0)
                args[length++] = ' ';

        memcpy(args + length, e, n);
        length += n;
}

if (!GREEDY_REALLOC(args, allocated, length + 1)
        return -ENOMEM;
args[length++] = 0;
…
```